File: databinding/src/main/java/com/ypw/viewmodel/ui/MainActivity.java
package com.ypw.viewmodel.ui;

import android.content.Intent;
import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import androidx.databinding.DataBindingUtil;
import com.ypw.viewmodel.R;
import com.ypw.viewmodel.databinding.ActivityMainBinding;


/**
 * databinding:通过页面和数据绑定展示数据
 * 数据绑定分为两种方式
 * 1.Observable:通过继承Observable;或者是持有实际对象的Observable来实现UI页面上数据的改变
 * 2.ViewModel+LiveData:
 */
public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        ActivityMainBinding binding = DataBindingUtil.setContentView(this, R.layout.activity_main);
        binding.observableActivityButton.setOnClickListener(v -> {
            startActivity(new Intent(this, ObservableActivity.class));
        });
        binding.viewmodelActivityButton.setOnClickListener(v -> {
            startActivity(new Intent(this, ViewModelActivity.class));
        });
    }
}

File: databinding_twoways/src/main/java/com/test/databinding_twoways/ui/MainActivity.kt
package com.test.databinding_twoways.ui

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.ViewModelProviders
import com.test.databinding_twoways.R
import com.test.databinding_twoways.data.IntervalTimerViewModel
import com.test.databinding_twoways.data.IntervalTimerViewModelFactory
import com.test.databinding_twoways.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {
    private val intervalTimerViewModel: IntervalTimerViewModel
            by lazy {
                ViewModelProviders.of(this, IntervalTimerViewModelFactory)
                    .get(IntervalTimerViewModel::class.java)
            }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val binding: ActivityMainBinding = DataBindingUtil.setContentView(this, R.layout.activity_main)
        val viewmodel = intervalTimerViewModel
        binding.viewmodel = viewmodel
    }
}
